libbpf-tools/readahead.c

```diff
@@ -119,28 +139,35 @@ int main(int argc, char **argv)
 		fprintf(stderr, "failed to open BPF object\n");
 		return 1;
 	}
-
-	/*
-	 * Starting from v5.10-rc1 (8238287), __do_page_cache_readahead has
-	 * renamed to do_page_cache_ra. So we specify the function dynamically.
-	 */
-	err = readahead__set_attach_target(obj->progs.do_page_cache_ra);
-	if (err)
-		goto cleanup;
-	err = readahead__set_attach_target(obj->progs.do_page_cache_ra_ret);
-	if (err)
-		goto cleanup;
-
 	err = readahead_bpf__load(obj);
```

Comment:
how did you test these changes? I suspect it won't work on some kernels, because for fentry/fexit programs you have to `bpf_program__set_autoload(false)` them on old kernels before `readahead_bpf__load()` is called, otherwise BPF verifier will reject them. Then, for cases where you do want to use fentry/fexit you need to do `readahead__set_attach_target()` before load as well to set correct target functions.

Lastly, for bpf_program__attach_kprobe, given you specify attach function name at the attach type, you don't need two copies of kretprobe and kprobe in BPF code, you can reuse the same function and just specify the right function name.

cc @ethercflow  I think given that now the conditions of attachment is more complicated, it is probably a good idea to invest in two helper method to check whether some function exists in kernel or not. How about we add few helpers:
  - `bool fentry_exists(const char *name, const char *mod);`. If `mod != NULL`, load `/sys/kernel/btf/<mod>` (if it exists) and check if FUNC with a given name exists; if not, load `/sys/kernel/btf/vmlinux` and check if FUNC with given name exists;
  - `bool kprobe_exists(const char *name);` which should scan `/sys/kernel/tracing/available_filter_functions` (probably with fallback to /proc/kallsyms, which is more expensive to parse). This can be used to detect `__do_page_cache_readahead` vs `do_page_cache_ra` for kprobe attachment case;
  
  So in the end, readahead logic would be something like:
```
if (fentry_exists("do_page_cache_ra", NULL)) {
    bpf_program__set_attach_target(obj->progs.do_page_cache_ra, "do_page_cache_ra");
  /* same for fexit */
} else if (fentry_exists("__do_page_cache_readahead")) {
    bpf_program__set_attach_target(obj->progs.do_page_cache_ra, "__do_page_cache_readahead");
} else if (kprobe_exists("do_page_cache_ra")) {
   /* use bpf_program__attach_kprobe(..., "do_page_cache_ra");
} else {
   bpf_program__attach_kprobe(..., "__do_page_cache_readahead")
}
```

Reply:
Yes, I agree. Compared with the previous way, the introduction of these two helpers will make the logic clearer.

Reply:
So for fentry_exists case, it's not enough to check that FUNC exists, we need to check that kernel actually supports fentry. As I pointed out in https://github.com/iovisor/bcc/pull/3214#discussion_r617058439, kernel BTF is exposed starting from 5.4 kernel, but fentry/fexit is actually supported starting from 5.5, so there is a gap between kernel versions. But good thing is that with BTF it's actually easy to detect if fentry/fexit is supported:

1. btf__find_by_name_kind(btf, "bpf_attach_type", BTF_KIND_ENUM).
2. If that enum has BPF_TRACE_FENTRY enum value, then kernel supports FENTRY programs.

@ethercflow, would you be able to contribute discussed helpers?

Reply:
@anakryiko Sure, I will do this in the next few days.